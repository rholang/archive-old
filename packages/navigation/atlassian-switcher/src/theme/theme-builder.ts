import memoizeOne from 'memoize-one';

import {
  ThemeProps,
  CustomThemeResult,
  ApplyThemeFn,
  ThemingPublicApi,
} from './types';

const maybeGetToken = (
  propertyName: string,
  customThemeProps: ThemingPublicApi,
  colorName: keyof ThemingPublicApi,
) => {
  const value = customThemeProps[colorName];

  if (!value) {
    return {};
  }

  return {
    [propertyName]: value,
  };
};
export const createCustomTheme = memoizeOne(
  (customThemeProps: ThemingPublicApi = {}): CustomThemeResult => {
    const topLevelItemWrapperTheme = (
      theme: ApplyThemeFn,
      props: ThemeProps,
    ) => ({
      ...theme(props),
      hover: {
        ...theme(props).hover,
        ...maybeGetToken(
          'background',
          customThemeProps,
          'secondaryHoverBackgroundColor',
        ),
      },
    });

    const itemTheme = (theme: ApplyThemeFn, props: ThemeProps) => ({
      ...theme(props),
      hover: {
        ...theme(props).hover,
        ...maybeGetToken(
          'background',
          customThemeProps,
          'primaryHoverBackgroundColor',
        ),
        ...maybeGetToken('text', customThemeProps, 'primaryTextColor'),
        ...maybeGetToken(
          'secondaryText',
          customThemeProps,
          'secondaryTextColor',
        ),
      },
      default: {
        ...theme(props).default,
        ...maybeGetToken('text', customThemeProps, 'primaryTextColor'),
        ...maybeGetToken(
          'secondaryText',
          customThemeProps,
          'secondaryTextColor',
        ),
      },
    });

    const childItemTheme = (theme: ApplyThemeFn, props: ThemeProps) => ({
      ...theme(props),
      hover: {
        ...theme(props).hover,
        ...maybeGetToken(
          'background',
          customThemeProps,
          'primaryHoverBackgroundColor',
        ),
        ...maybeGetToken('text', customThemeProps, 'primaryTextColor'),
        ...maybeGetToken(
          'secondaryText',
          customThemeProps,
          'secondaryTextColor',
        ),
      },
      default: {
        ...theme(props).default,
        ...maybeGetToken(
          'background',
          customThemeProps,
          'secondaryHoverBackgroundColor',
        ),
        ...maybeGetToken('text', customThemeProps, 'primaryTextColor'),
        ...maybeGetToken(
          'secondaryText',
          customThemeProps,
          'secondaryTextColor',
        ),
      },
    });

    return {
      topLevelItemWrapperTheme,
      itemTheme,
      childItemTheme,
    };
  },
);
