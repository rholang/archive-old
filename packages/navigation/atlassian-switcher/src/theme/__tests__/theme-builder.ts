import { createCustomTheme } from '../theme-builder';

describe('utils/theme-builder', () => {
  const noopThemeFn = () => ({});

  it('should export a theming function with Theme components for all the expected items', () => {
    expect(createCustomTheme({})).toHaveProperty('topLevelItemWrapperTheme');
    expect(createCustomTheme({})).toHaveProperty('itemTheme');
    expect(createCustomTheme({})).toHaveProperty('childItemTheme');
  });

  describe('creating a proper theme', () => {
    const blueColorScheme = {
      primaryTextColor: '#0000B2',
      secondaryTextColor: '#03396c',
      primaryHoverBackgroundColor: '#ccffff',
      secondaryHoverBackgroundColor: '#e5ffff',
    };

    const themes = createCustomTheme(blueColorScheme);

    it('topLevelItemWrapperTheme should match snapshot', () => {
      const itemTheme = themes.itemTheme(noopThemeFn, {});

      expect(itemTheme).toMatchSnapshot();
    });

    it('itemTheme should match snapshot', () => {
      const itemTheme = themes.itemTheme(noopThemeFn, {});

      expect(itemTheme).toMatchSnapshot();
    });

    it('childItemTheme should match snapshot', () => {
      const childItemTheme = themes.childItemTheme(noopThemeFn, {});

      expect(childItemTheme).toMatchSnapshot();
    });
  });

  describe('creating a theme without passing all the props will fallback to default', () => {
    const blueColorScheme = {
      primaryTextColor: '#0000B2',
      primaryHoverBackgroundColor: '#ccffff',
    };

    const themes = createCustomTheme(blueColorScheme);

    it('topLevelItemWrapperTheme should match snapshot', () => {
      const itemTheme = themes.itemTheme(noopThemeFn, {});

      expect(itemTheme).toMatchSnapshot();
    });

    it('itemTheme should match snapshot', () => {
      const itemTheme = themes.itemTheme(noopThemeFn, {});

      expect(itemTheme).toMatchSnapshot();
    });

    it('childItemTheme should match snapshot', () => {
      const childItemTheme = themes.childItemTheme(noopThemeFn, {});

      expect(childItemTheme).toMatchSnapshot();
    });
  });
});
