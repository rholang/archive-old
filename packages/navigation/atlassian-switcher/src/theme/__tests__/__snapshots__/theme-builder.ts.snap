// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/theme-builder creating a proper theme childItemTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "background": "#e5ffff",
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
}
`;

exports[`utils/theme-builder creating a proper theme itemTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
}
`;

exports[`utils/theme-builder creating a proper theme topLevelItemWrapperTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "secondaryText": "#03396c",
    "text": "#0000B2",
  },
}
`;

exports[`utils/theme-builder creating a theme without passing all the props will fallback to default childItemTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "text": "#0000B2",
  },
}
`;

exports[`utils/theme-builder creating a theme without passing all the props will fallback to default itemTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "text": "#0000B2",
  },
}
`;

exports[`utils/theme-builder creating a theme without passing all the props will fallback to default topLevelItemWrapperTheme should match snapshot 1`] = `
Object {
  "default": Object {
    "text": "#0000B2",
  },
  "hover": Object {
    "background": "#ccffff",
    "text": "#0000B2",
  },
}
`;
